File: clients/rust/turk-router/tests/common/fixture.rs
//! The fixture text format, parsed and rendered byte for byte.
//!
//! A fixture is one pool's window as the program's own test harness captured it: a header, then
//! one block per window slot after the venue program, then the mint accounts as extras. Slot
//! roles say what the harness substitutes at run time (`in_ata`, `out_ata`, `payer`) and what it
//! installs as captured (`fixed`, which carries a body) or references by address (`program_ref`).

use solana_pubkey::Pubkey;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Fixture {
    pub kind: String,
    pub pool_b58: String,
    pub slot: u64,
    pub unix_ts: i64,
    pub hop_kind: u8,
    pub program_id: Pubkey,
    pub input_mint: Pubkey,
    pub output_mint: Pubkey,
    pub input_token_program: Pubkey,
    pub output_token_program: Pubkey,
    pub hook_lens: [u8; 2],
    pub slots: Vec<Slot>,
    pub extras: Vec<Extra>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Role {
    Fixed,
    InAta,
    OutAta,
    Payer,
    ProgramRef,
}

impl Role {
    fn parse(text: &str) -> Role {
        match text {
            "fixed" => Role::Fixed,
            "in_ata" => Role::InAta,
            "out_ata" => Role::OutAta,
            "payer" => Role::Payer,
            "program_ref" => Role::ProgramRef,
            other => panic!("unknown role {other}"),
        }
    }

    fn name(self) -> &'static str {
        match self {
            Role::Fixed => "fixed",
            Role::InAta => "in_ata",
            Role::OutAta => "out_ata",
            Role::Payer => "payer",
            Role::ProgramRef => "program_ref",
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Body {
    pub owner: Pubkey,
    pub lamports: u64,
    pub data: Vec<u8>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Slot {
    pub writable: bool,
    pub signer: bool,
    pub role: Role,
    pub pubkey: Pubkey,
    pub body: Option<Body>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Extra {
    pub role: String,
    pub pubkey: Pubkey,
    pub body: Body,
}

impl Fixture {
    /// The window the program sees: the venue program, then every slot.
    pub fn window_pubkeys(&self) -> Vec<Pubkey> {
        std::iter::once(self.program_id)
            .chain(self.slots.iter().map(|slot| slot.pubkey))
            .collect()
    }

    pub fn slot(&self, index: usize) -> &Slot {
        self.slots
            .get(index)
            .unwrap_or_else(|| panic!("fixture has no slot {index}"))
    }

    pub fn extra(&self, role: &str) -> &Extra {
        self.extras
            .iter()
            .find(|extra| extra.role == role)
            .unwrap_or_else(|| panic!("fixture has no extra {role}"))
    }
}

enum Target {
    Header,
    Slot,
    Extra,
}

pub fn parse(text: &str) -> Fixture {
    let mut kind = None;
    let mut pool_b58 = None;
    let mut slot = None;
    let mut unix_ts = None;
    let mut hop_kind = None;
    let mut hook_lens = [0u8, 0u8];
    let mut program_id = None;
    let mut input_mint = None;
    let mut output_mint = None;
    let mut input_token_program = None;
    let mut output_token_program = None;
    let mut slots: Vec<Slot> = Vec::new();
    let mut extras: Vec<Extra> = Vec::new();

    let mut target = Target::Header;
    let mut pubkey: Option<Pubkey> = None;
    let mut owner: Option<Pubkey> = None;
    let mut lamports: Option<u64> = None;
    let mut data: Option<Vec<u8>> = None;

    fn flush(
        target: &Target,
        slots: &mut [Slot],
        extras: &mut [Extra],
        pubkey: &mut Option<Pubkey>,
        owner: &mut Option<Pubkey>,
        lamports: &mut Option<u64>,
        data: &mut Option<Vec<u8>>,
    ) {
        match target {
            Target::Header => {}
            Target::Slot => {
                let entry = slots.last_mut().expect("account header seen");
                entry.pubkey = pubkey.take().expect("account block has pubkey");
                entry.body = match (owner.take(), lamports.take(), data.take()) {
                    (Some(owner), Some(lamports), Some(data)) => Some(Body {
                        owner,
                        lamports,
                        data,
                    }),
                    (None, None, None) => None,
                    _ => panic!("partial account body"),
                };
            }
            Target::Extra => {
                let entry = extras.last_mut().expect("extra header seen");
                entry.pubkey = pubkey.take().expect("extra block has pubkey");
                entry.body = Body {
                    owner: owner.take().expect("extra has owner"),
                    lamports: lamports.take().expect("extra has lamports"),
                    data: data.take().expect("extra has data"),
                };
            }
        }
    }

    for line in text.lines() {
        if line.is_empty() {
            continue;
        }
        let (key, value) = match line.split_once(':') {
            Some((key, value)) => (key.trim(), value.trim()),
            None => (line, ""),
        };
        match key {
            "account" => {
                flush(
                    &target,
                    &mut slots,
                    &mut extras,
                    &mut pubkey,
                    &mut owner,
                    &mut lamports,
                    &mut data,
                );
                target = Target::Slot;
                slots.push(Slot {
                    writable: header_field(value, "writable=") == "true",
                    signer: header_field(value, "signer=") == "true",
                    role: Role::parse(header_field(value, "role=")),
                    pubkey: Pubkey::default(),
                    body: None,
                });
            }
            "extra" => {
                flush(
                    &target,
                    &mut slots,
                    &mut extras,
                    &mut pubkey,
                    &mut owner,
                    &mut lamports,
                    &mut data,
                );
                target = Target::Extra;
                extras.push(Extra {
                    role: header_field(value, "role=").to_string(),
                    pubkey: Pubkey::default(),
                    body: Body {
                        owner: Pubkey::default(),
                        lamports: 0,
                        data: Vec::new(),
                    },
                });
            }
            "pubkey" => pubkey = Some(parse_key(value)),
            "owner" => owner = Some(parse_key(value)),
            "lamports" => lamports = Some(value.parse().expect("lamports u64")),
            "data_hex" => data = Some(hex_decode(value)),
            "hop_kind" => hop_kind = Some(value.parse().expect("hop_kind u8")),
            "hook_lens" => {
                let (a, b) = value.split_once(',').expect("hook_lens is 'A,B'");
                hook_lens = [
                    a.trim().parse().expect("hook_lens[0] u8"),
                    b.trim().parse().expect("hook_lens[1] u8"),
                ];
            }
            "slot" => slot = Some(value.parse().expect("slot u64")),
            "unix_ts" => unix_ts = Some(value.parse().expect("unix_ts i64")),
            "program_id" => program_id = Some(parse_key(value)),
            "input_mint" => input_mint = Some(parse_key(value)),
            "output_mint" => output_mint = Some(parse_key(value)),
            "input_token_program" => input_token_program = Some(parse_key(value)),
            "output_token_program" => output_token_program = Some(parse_key(value)),
            "kind" => kind = Some(value.to_string()),
            "pool_b58" => pool_b58 = Some(value.to_string()),
            _ => {}
        }
    }
    flush(
        &target,
        &mut slots,
        &mut extras,
        &mut pubkey,
        &mut owner,
        &mut lamports,
        &mut data,
    );

    Fixture {
        kind: kind.expect("kind header"),
        pool_b58: pool_b58.expect("pool_b58 header"),
        slot: slot.expect("slot header"),
        unix_ts: unix_ts.expect("unix_ts header"),
        hop_kind: hop_kind.expect("hop_kind header"),
        program_id: program_id.expect("program_id header"),
        input_mint: input_mint.expect("input_mint header"),
        output_mint: output_mint.expect("output_mint header"),
        input_token_program: input_token_program.expect("input_token_program header"),
        output_token_program: output_token_program.expect("output_token_program header"),
        hook_lens,
        slots,
        extras,
    }
}

pub fn render(fixture: &Fixture) -> String {
    let mut out = String::new();
    out.push_str(&format!("kind: {}\n", fixture.kind));
    out.push_str(&format!("pool_b58: {}\n", fixture.pool_b58));
    out.push_str(&format!("slot: {}\n", fixture.slot));
    out.push_str(&format!("unix_ts: {}\n", fixture.unix_ts));
    out.push_str(&format!("hop_kind: {}\n", fixture.hop_kind));
    out.push_str(&format!(
        "program_id: {}\n",
        format_key(&fixture.program_id)
    ));
    out.push_str(&format!(
        "input_mint: {}\n",
        format_key(&fixture.input_mint)
    ));
    out.push_str(&format!(
        "output_mint: {}\n",
        format_key(&fixture.output_mint)
    ));
    out.push_str(&format!(
        "input_token_program: {}\n",
        format_key(&fixture.input_token_program)
    ));
    out.push_str(&format!(
        "output_token_program: {}\n",
        format_key(&fixture.output_token_program)
    ));
    out.push_str(&format!("accounts: {}\n", fixture.slots.len()));
    if fixture.hook_lens != [0, 0] {
        out.push_str(&format!(
            "hook_lens: {},{}\n",
            fixture.hook_lens[0], fixture.hook_lens[1]
        ));
    }
    for (index, slot) in fixture.slots.iter().enumerate() {
        out.push_str(&format!(
            "account: index={index} writable={} signer={} role={}\n",
            slot.writable,
            slot.signer,
            slot.role.name()
        ));
        out.push_str(&format!("pubkey: {}\n", format_key(&slot.pubkey)));
        if let Some(body) = &slot.body {
            render_body(&mut out, body);
        }
    }
    for extra in &fixture.extras {
        out.push_str(&format!("extra: role={}\n", extra.role));
        out.push_str(&format!("pubkey: {}\n", format_key(&extra.pubkey)));
        render_body(&mut out, &extra.body);
    }
    out
}

fn render_body(out: &mut String, body: &Body) {
    out.push_str(&format!("owner: {}\n", format_key(&body.owner)));
    out.push_str(&format!("lamports: {}\n", body.lamports));
    out.push_str(&format!("data_hex: {}\n", hex_encode(&body.data)));
}

fn header_field<'a>(header: &'a str, key: &str) -> &'a str {
    let from = header.find(key).expect("header field") + key.len();
    header[from..]
        .split_whitespace()
        .next()
        .expect("header field value")
}

fn parse_key(value: &str) -> Pubkey {
    let inner = value
        .strip_prefix('[')
        .and_then(|rest| rest.strip_suffix(']'))
        .expect("a bracketed key");
    let bytes: Vec<u8> = inner
        .split(',')
        .map(|token| token.trim().parse().expect("a byte"))
        .collect();
    Pubkey::new_from_array(<[u8; 32]>::try_from(bytes).expect("32 bytes"))
}

fn format_key(key: &Pubkey) -> String {
    let joined = key
        .to_bytes()
        .iter()
        .map(u8::to_string)
        .collect::<Vec<_>>()
        .join(", ");
    format!("[{joined}]")
}

fn hex_decode(text: &str) -> Vec<u8> {
    assert!(text.len().is_multiple_of(2), "odd-length hex");
    (0..text.len() / 2)
        .map(|index| u8::from_str_radix(&text[index * 2..index * 2 + 2], 16).expect("hex byte"))
        .collect()
}

fn hex_encode(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

File: clients/ts/tests/common/fixture.ts
/**
 * The fixture text format, parsed and rendered byte for byte — a port of the Rust test suite's
 * `tests/common/fixture.rs`, with the same grammar so a malformed file fails the same way on
 * both sides.
 */
import { getAddressDecoder, getAddressEncoder, type Address } from "@solana/addresses";
import { getBase16Decoder, getBase16Encoder } from "@solana/codecs-strings";

export type Role = "fixed" | "in_ata" | "out_ata" | "payer" | "program_ref";

const ROLES: ReadonlyArray<Role> = ["fixed", "in_ata", "out_ata", "payer", "program_ref"];

export interface Body {
  readonly owner: Address;
  readonly lamports: bigint;
  readonly data: Uint8Array;
}

export interface Slot {
  readonly writable: boolean;
  readonly signer: boolean;
  readonly role: Role;
  readonly pubkey: Address;
  readonly body: Body | undefined;
}

export interface Extra {
  readonly role: string;
  readonly pubkey: Address;
  readonly body: Body;
}

export interface Fixture {
  readonly kind: string;
  readonly poolB58: string;
  readonly slot: bigint;
  readonly unixTs: bigint;
  readonly hopKind: number;
  readonly programId: Address;
  readonly inputMint: Address;
  readonly outputMint: Address;
  readonly inputTokenProgram: Address;
  readonly outputTokenProgram: Address;
  readonly hookLens: readonly [number, number];
  readonly slots: Array<Slot>;
  readonly extras: Array<Extra>;
}

export function windowPubkeys(fixture: Fixture): Array<Address> {
  return [fixture.programId, ...fixture.slots.map((slot) => slot.pubkey)];
}

export function slotAt(fixture: Fixture, index: number): Slot {
  const slot = fixture.slots[index];
  if (slot === undefined) {
    throw new Error(`fixture has no slot ${index}`);
  }
  return slot;
}

interface PendingSlot {
  readonly kind: "slot";
  readonly writable: boolean;
  readonly signer: boolean;
  readonly role: Role;
}

interface PendingExtra {
  readonly kind: "extra";
  readonly role: string;
}

interface Pending {
  header: PendingSlot | PendingExtra | undefined;
  pubkey: Address | undefined;
  owner: Address | undefined;
  lamports: bigint | undefined;
  data: Uint8Array | undefined;
}

export function parse(text: string): Fixture {
  const header: Record<string, string> = {};
  let hookLens: readonly [number, number] = [0, 0];
  const slots: Array<Slot> = [];
  const extras: Array<Extra> = [];
  const pending: Pending = {
    header: undefined,
    pubkey: undefined,
    owner: undefined,
    lamports: undefined,
    data: undefined,
  };

  const flush = (): void => {
    const block = pending.header;
    if (block === undefined) {
      return;
    }
    if (pending.pubkey === undefined) {
      throw new Error("account block has no pubkey");
    }
    const bodyFields = [pending.owner, pending.lamports, pending.data];
    const present = bodyFields.filter((value) => value !== undefined).length;
    let body: Body | undefined;
    if (present === 3 && pending.owner !== undefined && pending.lamports !== undefined) {
      body = {
        owner: pending.owner,
        lamports: pending.lamports,
        data: pending.data ?? new Uint8Array(),
      };
    } else if (present !== 0) {
      throw new Error("partial account body");
    }
    if (block.kind === "slot") {
      slots.push({
        writable: block.writable,
        signer: block.signer,
        role: block.role,
        pubkey: pending.pubkey,
        body,
      });
    } else {
      if (body === undefined) {
        throw new Error("extra has no body");
      }
      extras.push({ role: block.role, pubkey: pending.pubkey, body });
    }
    pending.header = undefined;
    pending.pubkey = undefined;
    pending.owner = undefined;
    pending.lamports = undefined;
    pending.data = undefined;
  };

  for (const rawLine of text.split("\n")) {
    const line = rawLine.endsWith("\r") ? rawLine.slice(0, -1) : rawLine;
    if (line.length === 0) {
      continue;
    }
    const colon = line.indexOf(":");
    const key = colon === -1 ? line : line.slice(0, colon).trim();
    const value = colon === -1 ? "" : line.slice(colon + 1).trim();
    switch (key) {
      case "account":
        flush();
        pending.header = {
          kind: "slot",
          writable: headerField(value, "writable=") === "true",
          signer: headerField(value, "signer=") === "true",
          role: parseRole(headerField(value, "role=")),
        };
        break;
      case "extra":
        flush();
        pending.header = { kind: "extra", role: headerField(value, "role=") };
        break;
      case "pubkey":
        pending.pubkey = parseKey(value);
        break;
      case "owner":
        pending.owner = parseKey(value);
        break;
      case "lamports":
        pending.lamports = BigInt(value);
        break;
      case "data_hex":
        pending.data = hexDecode(value);
        break;
      case "hook_lens": {
        const [first, second] = value.split(",");
        if (first === undefined || second === undefined) {
          throw new Error("hook_lens is 'A,B'");
        }
        hookLens = [Number(first.trim()), Number(second.trim())];
        break;
      }
      case "kind":
      case "pool_b58":
      case "slot":
      case "unix_ts":
      case "hop_kind":
      case "program_id":
      case "input_mint":
      case "output_mint":
      case "input_token_program":
      case "output_token_program":
        header[key] = value;
        break;
      default:
        break;
    }
  }
  flush();

  return {
    kind: headerText(header, "kind"),
    poolB58: headerText(header, "pool_b58"),
    slot: BigInt(headerText(header, "slot")),
    unixTs: BigInt(headerText(header, "unix_ts")),
    hopKind: Number(headerText(header, "hop_kind")),
    programId: parseKey(headerText(header, "program_id")),
    inputMint: parseKey(headerText(header, "input_mint")),
    outputMint: parseKey(headerText(header, "output_mint")),
    inputTokenProgram: parseKey(headerText(header, "input_token_program")),
    outputTokenProgram: parseKey(headerText(header, "output_token_program")),
    hookLens,
    slots,
    extras,
  };
}

export function render(fixture: Fixture): string {
  let out = "";
  out += `kind: ${fixture.kind}\n`;
  out += `pool_b58: ${fixture.poolB58}\n`;
  out += `slot: ${String(fixture.slot)}\n`;
  out += `unix_ts: ${String(fixture.unixTs)}\n`;
  out += `hop_kind: ${String(fixture.hopKind)}\n`;
  out += `program_id: ${formatKey(fixture.programId)}\n`;
  out += `input_mint: ${formatKey(fixture.inputMint)}\n`;
  out += `output_mint: ${formatKey(fixture.outputMint)}\n`;
  out += `input_token_program: ${formatKey(fixture.inputTokenProgram)}\n`;
  out += `output_token_program: ${formatKey(fixture.outputTokenProgram)}\n`;
  out += `accounts: ${String(fixture.slots.length)}\n`;
  if (fixture.hookLens[0] !== 0 || fixture.hookLens[1] !== 0) {
    out += `hook_lens: ${String(fixture.hookLens[0])},${String(fixture.hookLens[1])}\n`;
  }
  fixture.slots.forEach((slot, index) => {
    out += `account: index=${String(index)} writable=${String(slot.writable)} signer=${String(slot.signer)} role=${slot.role}\n`;
    out += `pubkey: ${formatKey(slot.pubkey)}\n`;
    if (slot.body !== undefined) {
      out += renderBody(slot.body);
    }
  });
  for (const extra of fixture.extras) {
    out += `extra: role=${extra.role}\n`;
    out += `pubkey: ${formatKey(extra.pubkey)}\n`;
    out += renderBody(extra.body);
  }
  return out;
}

function renderBody(body: Body): string {
  return (
    `owner: ${formatKey(body.owner)}\n` +
    `lamports: ${String(body.lamports)}\n` +
    `data_hex: ${getBase16Decoder().decode(body.data)}\n`
  );
}

function headerText(header: Record<string, string>, key: string): string {
  const value = header[key];
  if (value === undefined) {
    throw new Error(`${key} header`);
  }
  return value;
}

function headerField(header: string, key: string): string {
  const from = header.indexOf(key);
  if (from === -1) {
    throw new Error(`header field ${key}`);
  }
  const rest = header.slice(from + key.length);
  const token = rest.split(/\s+/)[0];
  if (token === undefined || token.length === 0) {
    throw new Error(`header field value ${key}`);
  }
  return token;
}

function parseRole(text: string): Role {
  const role = ROLES.find((candidate) => candidate === text);
  if (role === undefined) {
    throw new Error(`unknown role ${text}`);
  }
  return role;
}

function parseKey(value: string): Address {
  if (!value.startsWith("[") || !value.endsWith("]")) {
    throw new Error("a bracketed key");
  }
  const bytes = value
    .slice(1, -1)
    .split(",")
    .map((token) => Number(token.trim()));
  if (
    bytes.length !== 32 ||
    bytes.some((byte) => !Number.isInteger(byte) || byte < 0 || byte > 255)
  ) {
    throw new Error("32 bytes");
  }
  return getAddressDecoder().decode(new Uint8Array(bytes));
}

function formatKey(address: Address): string {
  return `[${Array.from(getAddressEncoder().encode(address), (byte) => String(byte)).join(", ")}]`;
}

// Kit's base16 encoder silently truncates an odd-length string; the Rust parser asserts, so this
// one does too.
function hexDecode(text: string): Uint8Array {
  if (text.length % 2 !== 0) {
    throw new Error("odd-length hex");
  }
  return new Uint8Array(getBase16Encoder().encode(text));
}
